### PR TITLE
[refactor] Refactor Actor and Critic classes

### PR DESCRIPTION
**Pull request by ervteng**
### Proposed change(s)

Restructure ActorCritic and SeparateActorCritic into two separate classes. Both have parent class Actor (this makes sense, as ActorCritic `is-a` Actor). 

Also adds typing to a ton of stuff. 

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [x] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [x] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/policy/torch_policy.py [hunk 1]
```diff
@@ -69,18 +70,26 @@ def __init__(
...
-            act_type=behavior_spec.action_type,
-            act_size=self.act_size,
-            stream_names=reward_signal_names,
-            separate_critic=separate_critic
-            if separate_critic is not None
-            else self.use_continuous_act,
-            conditional_sigma=self.condition_sigma_on_obs,
-            tanh_squash=tanh_squash,
-        )
+        if separate_critic:
```
Comment (awjuliani) [inline ml-agents/mlagents/trainers/policy/torch_policy.py:73]: I think we can grab the right class constructor in the if/then statement, and then use it to instantiate the class after. That way we save a bunch of lines.

Comment (awjuliani) [reply ml-agents/mlagents/trainers/policy/torch_policy.py:73]:
````
if separate_critic:
    ac_class = SeparateActorCritic
else:
    ac_class = ActorCritic

Comment (ervteng) [reply ml-agents/mlagents/trainers/policy/torch_policy.py:73]: Done

File: ml-agents/mlagents/trainers/torch/encoders.py [hunk 2]
```diff
@@ -180,10 +182,12 @@ def __init__(self, height, width, initial_channels, output_size):
         self.dense = nn.Linear(self.final_flat, self.h_size)
 
     def forward(self, visual_obs):
-        conv_1 = torch.relu(self.conv1(visual_obs))
-        conv_2 = torch.relu(self.conv2(conv_1))
-        conv_3 = torch.relu(self.conv3(conv_2))
-        hidden = torch.relu(self.dense(conv_3.view([-1, self.final_flat])))
+        conv_1 = nn.functional.leaky_relu(self.conv1(visual_obs))
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/encoders.py:185]: Arn't we using swish in the TensorFlow code ?

File: ml-agents/mlagents/trainers/torch/networks.py [hunk 3]
```diff
@@ -262,6 +240,142 @@ def forward(
...
+        if self.act_type == ActionType.CONTINUOUS:
+            dists = self.distribution(encoding)
+        else:
+            dists = self.distribution(encoding, masks=masks)
+
+        value_outputs = self.value_heads(encoding)
+        return dists, value_outputs, memories
+
+
+class SeparateActorCritic(Actor):
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/networks.py:294]: Should all these actorcritic classes also inherit from a base Critic?

File: ml-agents/mlagents/trainers/torch/networks.py [hunk 4]
```diff
@@ -137,22 +139,20 @@ def forward(
...
+        encoding, memories = self.network_body(
             vec_inputs, vis_inputs, actions, memories, sequence_length
         )
-        output, _ = self.value_heads(embedding)
+        output = self.value_heads(encoding)
         return output, memories
 
 
-class ActorCritic(nn.Module):
+class Actor(nn.Module):
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/networks.py:149]: I think Actor should be an interface (not inherit from nn.Module and only have abstract methods). Then have a "SimpleActor" if we need a base implementation. I think that would make the API on the Actor much clearer

Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/networks.py:149]: Would you still do `Actor -> SimpleActor -> ActorCritic -> SeparateActorCritic`?

File: ml-agents/mlagents/trainers/torch/networks.py (hunk 3 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/networks.py:294]: I can do a base ActorCritic and then the other two being CombinedActorCritic and SeparateActorCritic. 

IMO the API for ActorCritic wouldn't make sense for a Critic, e.g. I'd expect the Critic's forward pass to be just the value heads while the ActorCritic should return actions.

File: ml-agents/mlagents/trainers/torch/encoders.py (hunk 2 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/encoders.py:185]: Yes - this is just to get the tests to pass (lots of zero gradients if you use ReLU). Will do a separate PR to change everything to Swish (requires some refactor of `utils.py` to avoid a circular import in the `encoders.py`.)

File: ml-agents/mlagents/trainers/torch/networks.py (hunk 4 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/networks.py]: Made some changes - it's now `SimpleActor(nn.Module, Actor)` and the other two inherit from that.

File: ml-agents/mlagents/trainers/torch/networks.py (hunk 3 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/networks.py]: Right now the ActorCritic class is using Duck typing 🦆

Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/networks.py]: Changed to using `ActorCritic` as an interface, and a `SharedActorCritic` and `SeparateActorCritic` classes.
